From 70c16bb518b3577ac0a86592cfbe4040b31c9ca6 Mon Sep 17 00:00:00 2001
From: "pi-cheng.chen" <pi-cheng.chen@linaro.org>
Date: Thu, 16 Apr 2015 17:26:07 +0800
Subject: [PATCH 3/3] comment out unworkable OPPs

---
 drivers/cpufreq/mt8173-cpufreq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/cpufreq/mt8173-cpufreq.c b/drivers/cpufreq/mt8173-cpufreq.c
index 1ab8330..e31b7e7 100644
--- a/drivers/cpufreq/mt8173-cpufreq.c
+++ b/drivers/cpufreq/mt8173-cpufreq.c
@@ -79,7 +79,7 @@ struct mtk_cpu_opp mt8173_l_opp[] = {
 	OPP(1183000000, 1028000, 0),
 	OPP(1404000000, 1083000, 0),
 	OPP(1508000000, 1109000, 0),
-	OPP(1573000000, 1125000, 0),
+/*	OPP(1573000000, 1125000, 0), */
 };
 
 /* OPP table for big cores of MT8173 */
@@ -91,7 +91,7 @@ struct mtk_cpu_opp mt8173_b_opp[] = {
 	OPP(1404000000, 1007000, 1107000),
 	OPP(1612000000, 1049000, 1149000),
 	OPP(1807000000, 1089000, 1150000),
-	OPP(1989000000, 1125000, 1150000),
+/*	OPP(1989000000, 1125000, 1150000), */
 };
 
 static inline int need_voltage_trace(struct cpu_dvfs_info *info)
-- 
1.9.1

