From e316d49eb26804cf9677a141b343a6b4a74c5218 Mon Sep 17 00:00:00 2001
From: Howard Chen <howard.chen@linaro.org>
Date: Mon, 23 Mar 2015 17:20:12 +0800
Subject: [PATCH 2/5] scp p1

---
 drivers/clk/mediatek/clk-mt8173.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/clk/mediatek/clk-mt8173.c b/drivers/clk/mediatek/clk-mt8173.c
index d98d88b..a9d95cb 100644
--- a/drivers/clk/mediatek/clk-mt8173.c
+++ b/drivers/clk/mediatek/clk-mt8173.c
@@ -25,6 +25,12 @@
 
 static DEFINE_SPINLOCK(mt8173_clk_lock);
 
+/* dirty hack to make scpsys work */
+spinlock_t *get_mtk_clk_lock(void)
+{
+	return &mt8173_clk_lock;
+}
+
 static const struct mtk_fixed_factor root_clk_alias[] __initconst = {
 	FACTOR(CLK_TOP_CLKPH_MCK_O, "clkph_mck_o", "clk_null", 1, 1),
 	FACTOR(CLK_TOP_DPI, "dpi_ck", "clk_null", 1, 1),
-- 
1.9.1

