From 6aaa430721b95897c21dd18bd6a24f6994b322d9 Mon Sep 17 00:00:00 2001
From: Howard Chen <howard.chen@linaro.org>
Date: Mon, 23 Mar 2015 17:20:22 +0800
Subject: [PATCH 3/5] scp p2

---
 drivers/clk/mediatek/clk-mtk.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/clk/mediatek/clk-mtk.h b/drivers/clk/mediatek/clk-mtk.h
index 61035b9..e9521ee 100644
--- a/drivers/clk/mediatek/clk-mtk.h
+++ b/drivers/clk/mediatek/clk-mtk.h
@@ -166,4 +166,11 @@ static inline void mtk_register_reset_controller(struct device_node *np,
 }
 #endif
 
+/* dirty hack to make scpsys work */
+extern spinlock_t *get_mtk_clk_lock(void);
+
+#define mtk_clk_lock(flags)    spin_lock_irqsave(get_mtk_clk_lock(), flags)
+#define mtk_clk_unlock(flags)  \
+               spin_unlock_irqrestore(get_mtk_clk_lock(), flags)
+
 #endif /* __DRV_CLK_MTK_H */
-- 
1.9.1

