From 1208579acdca6c740a18463cd81a5d98d8a5d7d6 Mon Sep 17 00:00:00 2001
From: Howard Chen <howard.chen@linaro.org>
Date: Mon, 23 Mar 2015 17:20:32 +0800
Subject: [PATCH 4/5] scp p3

---
 include/dt-bindings/clock/mt8173-clk.h | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/include/dt-bindings/clock/mt8173-clk.h b/include/dt-bindings/clock/mt8173-clk.h
index a9d682d..51da204 100644
--- a/include/dt-bindings/clock/mt8173-clk.h
+++ b/include/dt-bindings/clock/mt8173-clk.h
@@ -213,4 +213,18 @@
 #define PERI_I2C6		35
 #define PERI_NR_CLK		36
 
+/* SCP_SYS */
+
+#define SCP_SYS_VDE            1
+#define SCP_SYS_MFG            2
+#define SCP_SYS_VEN            3
+#define SCP_SYS_ISP            4
+#define SCP_SYS_DIS            5
+#define SCP_SYS_VEN2           6
+#define SCP_SYS_AUDIO          7
+#define SCP_SYS_MFG_2D         8
+#define SCP_SYS_MFG_ASYNC      9
+#define SCP_SYS_USB            10
+#define SCP_NR_CLK             11
+
 #endif /* _DT_BINDINGS_CLK_MT8173_H */
-- 
1.9.1

