From 42c82ca4e11d5a90cca9867b8df741473e936a89 Mon Sep 17 00:00:00 2001
From: Howard Chen <howard.chen@linaro.org>
Date: Mon, 23 Mar 2015 17:20:46 +0800
Subject: [PATCH 5/5] scp p4

---
 drivers/clk/mediatek/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/clk/mediatek/Makefile b/drivers/clk/mediatek/Makefile
index 299917a..cab01ce 100644
--- a/drivers/clk/mediatek/Makefile
+++ b/drivers/clk/mediatek/Makefile
@@ -1,4 +1,4 @@
 obj-y += clk-mtk.o clk-pll.o clk-gate.o clk-cpumux.o
 obj-$(CONFIG_RESET_CONTROLLER) += reset.o
 obj-y += clk-mt8135.o
-obj-y += clk-mt8173.o
+obj-y += clk-mt8173.o clk-mt8173-pg.o
-- 
1.9.1

