From 52e8bf90f14a976ddd4e0cdfdf2730073c69002a Mon Sep 17 00:00:00 2001
From: "freedom.tan" <freedom.tan@mediatek.com>
Date: Thu, 21 May 2015 08:57:25 +0800
Subject: [PATCH 3/3] idle workaround

---
 drivers/cpuidle/cpuidle-arm.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/cpuidle/cpuidle-arm.c b/drivers/cpuidle/cpuidle-arm.c
index 545069d..9ec73a3 100644
--- a/drivers/cpuidle/cpuidle-arm.c
+++ b/drivers/cpuidle/cpuidle-arm.c
@@ -50,7 +50,10 @@ static int arm_enter_idle_state(struct cpuidle_device *dev,
 		 * call the CPU ops suspend protocol with idle index as a
 		 * parameter.
 		 */
-		arm_cpuidle_suspend(idx);
+		// arm_cpuidle_suspend(idx);
+		// workaround for now
+		cpu_do_idle();
+		ret = idx;
 
 		cpu_pm_exit();
 	}
-- 
1.9.1

