From 9f016a8951c7130c34c899d772c2b844baccfd69 Mon Sep 17 00:00:00 2001
From: Howard Chen <howard.chen@linaro.org>
Date: Fri, 20 Mar 2015 12:19:19 +0800
Subject: [PATCH 12/14] tmp/drivers_usb_phy_Makefile.p1

---
 drivers/usb/phy/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/usb/phy/Makefile b/drivers/usb/phy/Makefile
index 75f2bba..d04246e 100644
--- a/drivers/usb/phy/Makefile
+++ b/drivers/usb/phy/Makefile
@@ -27,3 +27,4 @@ obj-$(CONFIG_USB_RCAR_GEN2_PHY)		+= phy-rcar-gen2-usb.o
 obj-$(CONFIG_USB_ULPI)			+= phy-ulpi.o
 obj-$(CONFIG_USB_ULPI_VIEWPORT)		+= phy-ulpi-viewport.o
 obj-$(CONFIG_KEYSTONE_USB_PHY)		+= phy-keystone.o
+obj-$(CONFIG_MT65XX_USB3PHY)		+= phy-mt65xx-usb3.o
-- 
1.9.1

